cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -268,79 +284,431 @@ func copyFromGlobHelper(src, dst, globEx string) error {
+	// SaveState is the state of the ubuntu-save mountpoint, it can be in any of
+	// the states documented for DataState, with the additional state of:
+	// - "not-found" for state when we don't have ubuntu-save, but it is
+	//   optional (unencrypted only)
+	SaveState string `json:"save-state"`
+	// HostLocation is the location where the host's ubuntu-data is mounted
+	// and available. If this is the empty string, then the host's
+	// ubuntu-data is not available anywhere.
+	HostLocation string `json:"host-location"`
+	// ErrorLog holds the log of errors that led to the degraded recovery mode.
+	ErrorLog []string `json:"error-log"`
+}
 
-	// get the disk that we mounted the ubuntu-seed partition from as a
-	// reference point for future mounts
-	disk, err := disks.DiskFromMountPoint(boot.InitramfsUbuntuSeedDir, nil)
+func (r *recoverDegradedState) LogErrorf(format string, v ...interface{}) {
+	msg := fmt.Sprintf(format, v...)
+	r.ErrorLog = append(r.ErrorLog, msg)
+	logger.Noticef(msg)
+}
+
+type recoverContext struct {
+	// the disk we have all our partitions on
+	disk disks.Disk
+	// options for finding partitions on the disk we have all our partitions on
+	diskOpts *disks.Options
+	// the device for ubuntu-data that is to be mounted - either an unencrypted
+	// volume, or a decrypted mapper volume if ubuntu-data is really encrypted
+	// on the physical disk
+	dataDevice string
+	// the device for ubuntu-save that is to be mounted
+	saveDevice string
+
+	// XXX: would be nice to get rid of tracking this here
+	isEncryptedDev bool
+
+	// state for tracking what happens
+	degradedState recoverDegradedState
+}
+
+// TODO: should this be a method on *context ?
+func verifyMountPointCtx(ctx *recoverContext, dir, name string) error {
+	matches, err := ctx.disk.MountPointIsFromDisk(dir, ctx.diskOpts)
 	if err != nil {
 		return err
 	}
+	if !matches {
+		return fmt.Errorf("cannot validate boot: %s mountpoint is expected to be from disk %s but is not", name, ctx.disk.Dev())
+	}
+	return nil
+}
 
-	// 3. mount ubuntu-data for recovery using run mode key
-	const lockKeysOnFinish = true
+// stateFunc is a function which executes a state action, returns the next
+// function (for the next) state or nil if it is the final state.
+type stateFunc func(ctx *recoverContext) (stateFunc, error)
+
+// stateMachine is a state machine implementing the logic for degraded recover
+// mode. the following state diagram shows the logic for the various states and
+// transitions:
+/**
+
+     fail or       +-------------------+  fail,     +-----------+fail,       +-------------+
+     not needed    |    locate save    |  unencrypt |unlock data|encrypted   | unlock data |
+    +--------------+    unencrypted    +<-----------+w/ run key +----------->+ w/ fallback +------+
+    |              |                   |            |           |            | key         |      |
+    |              +--------+----------+            +-----+-----+            +-----+-------+      |
```

Comment:
so we should try this then?
- unlock data with run key
- (if previous failed) unlock data with fallback key
- (if previous failed) unlock data with recovery key (asks for user input)
- proceed to save

Reply:
this logic should be implemented now